src/harvester/client.rs

```diff
@@ -0,0 +1,180 @@
+use std::env::var;
+use std::fmt;
+use std::future::Future;
+use std::sync::Arc;
+
+use anyhow::{Error, Result};
+use async_compression::tokio::{bufread::ZstdDecoder, write::ZstdEncoder};
+use bytes::Bytes;
+use cap_std::fs::Dir;
+use reqwest::Client as HttpClient;
+use tokio::time::{sleep, Duration};
+use tokio::{
+    fs::File as AsyncFile,
+    io::{AsyncReadExt, AsyncWriteExt, BufReader},
+};
+
+#[derive(Clone)]
+pub struct Client {
+    http_client: Option<HttpClient>,
+    dir: Arc<Dir>,
+}
+
+impl Client {
+    pub fn start(dir: &Dir) -> Result<Self> {
+        let http_client = match var("REPLAY_RESPONSES") {
+            Err(_err) => {
+                let http_client = HttpClient::builder()
+                    .user_agent("umwelt.info harvester")
+                    .timeout(Duration::from_secs(300))
+                    .build()?;
+
+                let _ = dir.remove_dir_all("responses");
+                dir.create_dir("responses")?;
+
+                Some(http_client)
+            }
+            Ok(_val) => None,
+        };
+
+        let dir = Arc::new(dir.open_dir("responses")?);
+
+        Ok(Self { dir, http_client })
+    }
+
+    pub async fn make_request<'a, A, F, T, E>(&'a self, key: &str, mut action: A) -> Result<T>
+    where
+        A: FnMut(&'a HttpClient) -> F,
+        F: Future<Output = Result<T, E>>,
+        T: Response,
+        E: Into<Error> + fmt::Display,
+    {
+        match &self.http_client {
+            Some(http_client) => {
+                let response = retry_request(|| action(http_client)).await?;
+
+                let file = self.dir.create(key)?;
+
+                let mut file = ZstdEncoder::new(AsyncFile::from_std(file.into_std()));
+                file.write_all(response.as_ref()).await?;
```

Comment:
Is this a good standard behavior to per default write all responses to disk? I guess for development this makes sense, but maybe disable it for "production" builds?

Reply:
I thought about this as well, but I see it as a debugging feature. And there the general wisdom is that debugging features belong in production builds because the bugs that need debugging usually happen in production. :-) (Well, there are plenty of bugs during development and testing, but they do not impact anything.)

So, since our service is network-bound (10% CPU utilization) and the disk footprint is manageable (50 MB), I would say we keep the function enabled until there are problems to be able to better diagnose harvester errors in production?

(This is of course only possible because this is not personally identifiable information, otherwise storing it would not fly.)

Reply:
I agree that we can keep it for the moment as it most certainly will help debugging. We should simply monitor how this develops if we have significantly more sources.